### DISCUSSION TITLE
Create new folder in FileExplorer

### DISCUSSION THREAD

**Discussion by lohi212** [codesandbox/sandpack#464] [Q&A]
Hey, I was using the FileExplorer as mentioned here: https://codesandbox.io/s/file-explorer-1ffn6?file=/src/App.tsx:547-565, https://github.com/codesandbox/sandpack/issues/98.

Is adding a new Folder possible on top of this?

**Comment by danilowoz** [comment]
Hey @lohi212! To be honest, we've never wondered about the need for this possibility. Would you mind describing how you would take advantage of this feature? What are the use cases?

Thanks, looking forward to helping you with this.

**Reply by theashishmaurya** [reply]
I think it will give a good coding environment for people who are building interactive Codelabs or something :)

**Reply by heykvr** [reply]
did they implemented that feature ?

**Comment by bilalahmed2500sxs** [comment]
hi @danilowoz , did code sandbox feature SandpackFileExplorer , we can add more files or not.

**Reply by danilowoz** [reply]
Hey, I'd recommend reading [this issue](https://github.com/codesandbox/sandpack/issues/913). Thanks for pinging me. I'll make sure to make it more transparent in the documentation.

**Comment by bilalahmed2500sxs** [comment]
Hi ma'am, I am facing a problem. I am using sanpack packages to create a
react environment. I built an react online editor for my application where
code is written and then stored in the database. When the code is stored in
the database then the react editor disappears. Now when I retrieve(code)
from the database I want it to show me a widget instead of code. Like that
i write a code for making buttons it will show button widget instead of
code. I have got an api and everything works smoothly but the data(Code)
present in the database does not convert it to a widget.I am very thankful
for that If you help me.

**Reply by lohi212** [reply]
Hi @bilalahmed2500sxs, Can you please provide a codesandbox link or some example with your use case?

**Reply by bilalahmed2500sxs** [reply]
Sure, first I will explain my application behavior.
 In my application there is a React editor. i want that  when a user writes code in react editor then it  clicks on the save button after clicking on the save button its code stored in the database and gives a widget .The user uses this widget by drag and drop functionality. i want to make my project (like Hostinger) where react is written and in response to it give me a widget.
Now my problem is that I saved my code in the database by making a react environment but in return did not give me a widget.
here is my react editor environment 
import React from "react";
import { CustomAceEditor } from "./CustomAceEditor";
import { SandpackProvider, SandpackPreview } from "@codesandbox/sandpack-react";

export const CustomSandpack = () => (
  <SandpackProvider template="react">
    <CustomAceEditor />
    <SandpackPreview showRefreshButton={false} showOpenInCodeSandbox={false} />
  </SandpackProvider>
);
and here is my CustomAceEditor 



import React from "react";
import axios from "axios";
import AceEditor from "react-ace";
import { useActiveCode } from "@codesandbox/sandpack-react";

import "ace-builds/src-noconflict/mode-javascript";
import "ace-builds/src-noconflict/theme-textmate";

export const CustomAceEditor = () => {
  const { code, updateCode } = useActiveCode();

  const handleSaveCode = async () => {
    try {
      await axios.post("http://localhost:3005/api/code", { code });
      alert("Code snippet saved successfully");
    } catch (error) {
      console.error("Error:", error);
      alert("An error occurred while saving the code snippet");
    }
    console.log("Code :", code);
  };

  return (
    <div>
      <AceEditor
        mode="javascript"
        defaultValue={code}
        onChange={updateCode}
        fontSize={14}
        height="300px"
        width="100%"
        setOptions={{ useWorker: false }}
      />
      <button onClick={handleSaveCode}>Save Code</button>
    </div>
  );
};

**Reply by lohi212** [reply]
Hey, You can look at this issue https://github.com/codesandbox/sandpack/discussions/314 and also check this CSB https://codesandbox.io/s/test-sandpack-code-k68y93?file=/CustomAceEditor.js.

**Reply by bilalahmed2500sxs** [reply]
@lohi212 thanks for your response first. but I explain above that after clicking on the save button its code stored in the database. now i want that code stored in the database give me widgets, fetched data from backend by get api, then using eval(), VM, and other transpilled code that convert the code present in the database give me widget , but right now its look impossible for me to do that . if you have any idea or you or your colleagues help me in that problem i am thankful for that.

**Comment by bilalahmed2500sxs** [comment]
Sure, first I will explain my application behaviour.
 In my application there is a React editor. i want that  when a user writes
code in react editor then it  clicks on the save button after clicking on
the save button its code stored in the database and gives a widget .The
user uses this widget by drag and drop functionality. i want to make my
project (like Hostinger) where react is written and in response to it give
me a widget.
Now my problem is that I saved my code in the database by making a react
environment but in return did not give me a widget.
here is my react editor environment
import React from "react";
import { CustomAceEditor } from "./CustomAceEditor";
import { SandpackProvider, SandpackPreview } from
***@***.***/sandpack-react";

export const CustomSandpack = () => (
  <SandpackProvider template="react">
    <CustomAceEditor />
    <SandpackPreview showRefreshButton={false} showOpenInCodeSandbox={false}
/>
  </SandpackProvider>
);
and here is my CustomAceEditor
import React from "react";
import axios from "axios";
import AceEditor from "react-ace";
import { useActiveCode } from ***@***.***/sandpack-react";

import "ace-builds/src-noconflict/mode-javascript";
import "ace-builds/src-noconflict/theme-textmate";

export const CustomAceEditor = () => {
  const { code, updateCode } = useActiveCode();

  const handleSaveCode = async () => {
    try {
      await axios.post("http://localhost:3005/api/code", { code });
      alert("Code snippet saved successfully");
    } catch (error) {
      console.error("Error:", error);
      alert("An error occurred while saving the code snippet");
    }
    console.log("Code :", code);
  };

  return (
    <div>
      <AceEditor
        mode="javascript"
        defaultValue={code}
        onChange={updateCode}
        fontSize={14}
        height="300px"
        width="100%"
        setOptions={{ useWorker: false }}
      />
      <button onClick={handleSaveCode}>Save Code</button>
    </div>
  );
};

**Comment by bilalahmed2500sxs** [comment]
@danilowoz hi, will we stored both code and its output in the database by using sandpack packages?

**Reply by danilowoz** [reply]
Hi, we don't store any code, dependency or even user usage on our database. Everything runs on the user's browser, and all public dependencies come from public CDN, such as jsdelivr.com

**Comment by bilalahmed2500sxs** [comment]
@lohi212  @danilowoz 
import {
  SandpackProvider,
  SandpackThemeProvider,
  SandpackPreview,
  SandpackCodeEditor,
  SandpackStack,
  SandpackLayout,
} from '@codesandbox/sandpack-react';
import { nightOwl } from '@codesandbox/sandpack-themes';
import { SandpackFileExplorer } from 'sandpack-file-explorer';

const MySandpackComponent = () => {
  return (
    <SandpackProvider template="react-ts">
      <SandpackThemeProvider theme={nightOwl}>
        <SandpackStack>
          <SandpackLayout>
            <div
              style={{
                display: 'flex',
                width: '100%',
                minHeight: '300px',
                maxHeight: '300px',
                backgroundColor: `var(--sp-colors-surface1)`,
              }}
            >
              <div
                style={{
                  minWidth: 150,
                  maxWidth: '300px',
                  overflow: 'hidden',
                }}
              >
                <SandpackFileExplorer />
              </div>
              <div style={{ flex: 'min-content' }}>
                <SandpackCodeEditor
                  wrapContent
                  style={{
                    minHeight: '100%',
                    maxHeight: '100%',
                    overflow: 'auto',
                  }}
                  showTabs
                  closableTabs
                  showInlineErrors
                  showLineNumbers
                />
              </div>
            </div>
            <SandpackPreview />
          </SandpackLayout>
        </SandpackStack>
      </SandpackThemeProvider>
    </SandpackProvider>
  );
};

will we use CustomAceEditor with it or not, instead of 
<SandpackCodeEditor
                  wrapContent
                  style={{
                    minHeight: '100%',
                    maxHeight: '100%',
                    overflow: 'auto',
                  }}
                  showTabs
                  closableTabs
                  showInlineErrors
                  showLineNumbers
                />
 </div>
